### DISCUSSION TITLE
Only support national highway/road networks that use prefixed country code network tags

### DISCUSSION THREAD

**Discussion by ZeLonewolf** [osm-americana/openstreetmap-americana#149] [General]
I propose that the project only support international road networks, as well as national and sub-national networks that are prefixed as described in the [Key:network](https://wiki.openstreetmap.org/wiki/Key:network#Highway_routes) wiki page.  As the leading community-developed highway shield renderer, we should promote the use of standards-based approaches to tagging and consuming route relation data, and avoid creating special cases unless it is unavoidable due to the complexity of the real-world route scheme.  We have a responsibility to the project as a whole to provide meaningful mapper feedback, and to not implement unreasonable workarounds to accommodate route tagging schemes that are inconsistent with the preponderance of global usage.

Encouraging mappers to embrace tagging schemes that better support route relation data consumers should be a goal of this project, even if it means delaying the rendering of a particular route network on the map.

This ticket is intended to capture a community discussion on whether we support a stricter approach towards inconsistent tagging approaches, or a more pragmatic approach that accepts developing special case code whenever needed to bridge inconsistent tagging schemes in order to maximize shield drawing.

**Comment by Mashin6** [comment]
The only caveat in this situation is that this system has never been formally proposed or approved. So for some people it might have even less credibility than standard tag that we vote on. 

But generally I agree with this approach. (_We just have to be careful, because there is a thin line between promoting good tagging and being a judge of what gets rendered ala OSM Carto._)

**Reply by 1ec5** [reply]
Americana doesn’t have an obligation to support everything right off the bat, so it can justifiably avoid supporting a particular tag value that’s problematic for technical reasons. Unlike an editor, a data consumer can do that without it necessarily being construed as promoting a particular viewpoint on tags. However, it does put facts on the ground that are relevant for any future debate about the tags. In this case, the technical challenge is that an unnamespaced `network` value can easily collide with an unrelated value. `network` was adopted for road route relations to avoid the messy situation of parsing the mini-language that had developed inside way `ref`s, but unnamespaced `network`s undercut that deliberate move.

In any case, many of the older network values were discussed extensively on mailing lists, forums, and wiki talk pages, just not put to a formal global vote on the wiki, being inherently regional. At this point, prefixing `network=*` with the country code is part of OSM’s “[unwritten constitution](https://en.wikipedia.org/wiki/Constitution_of_the_United_Kingdom)”. Of [the early documented networks](https://wiki.openstreetmap.org/wiki/Special:PermanentLink/377979#Road_Routes), `BAB` for German Autobahnen and `e-road` for the E-road network are the only values that aren’t prefixed with country codes. The E-road network is special, leaving Germany as the outlier.

**Reply by ZeLonewolf** [reply]
Switzerland is also an outlier.  They use `network=motorway`.

**Comment by coolultra1** [comment]
As vector tiles slowly approach reality and stuff like regional shields becomes possible, tagging will have to be more standardised, even if it means mass edits that might seem pointless to the bare eye.

**Comment by zekefarwell** [comment]
I think this is a good idea.  I'd like to see this project promote a standard tagging scheme as much as possible.  I also notice that the country codes are sometimes capitalized and sometimes not.  We could:

1. Only support the capitalized version that the wiki suggests is preferred
2. Support case insensitive matching (either capital or lowercase is fine in all cases)
3. Only support the dominant casing pattern in each country

Any thoughts on this?

**Reply by Mashin6** [reply]
I really like neatness and consistency so I would go with option 1. But we have to consider if enough countries would pass this filter. If there are too few shields rendered people might not be so motivated to add their country.

In Europe
Currently passing  1.: Austria, Czechia, Italy, Hungary, Slovenia, Serbia, Netherlands, Greece, France, Kosovo, Spain, Portugal, Montenegro, Romania
Passing 2.: Bulgaria, Bosnia and Herzegovina, Poland, Lithuania, Latvia, Estonia, Finland, Denmark, Ukraine, Belarus, Russia, Moldova, North Macedonia, Slovakia

Not passing 1 and 2: e-roads, Germany, Switzerland, Britain, Ireland, Iceland, Croatia

**Reply by 1ec5** [reply]
As I see it, the main practical problem is the lack of a country prefix on `network=BAB`. If a country or region is internally inconsistent regarding its `network` formats, that would be a problem too. Otherwise, the inconsistencies in capitalization or punctuation are cosmetic. There’s little point in equating a `*:primary` network of one country with a `*:Primary` network of another country anyways. Essentially, a data consumer can choose to be opinionated, but it should also pick its battles, so to speak.

**Reply by ZeLonewolf** [reply]
I would take a slightly more nuanced approach - if both upper and lowercase are in use in a country, we should use the uppercase only.  If only the lowercase is in use across an entire country, I think it's okay use that.  I think the important thing is that there's a unique prefix that's in use for that country's road system that we can rely upon as a distinguishing indicator.

**Reply by ZeLonewolf** [reply]
I'm also okay with international routes not having a prefix, after all they're not in a country.

**Comment by lonvia** [comment]
So what about other data users that rely on a long established standard like `network=BAB`? Please keep in mind that you are a relative late-comer in the rendering game. Something like 10 years late. As such I'd be really, really careful about dictating new standards.

**Reply by ZeLonewolf** [reply]
It would be helpful to understand whether there are specific examples of this, or whether this is just a hypothetical.

**Reply by lonvia** [reply]
Isn't it rather your task to prove that there are no examples before you propose such changes? The wiki clearly states: "In **some** countries, network values for highway routes follow a hierarchical scheme." If Germany is not in the group of the some, so be it. You are not promoting any standardized use here. You propose to change the tagging. No less.

**Reply by ZeLonewolf** [reply]
I propose no such thing.  As renderer authors, it is within our purview to decide which features and tagging we choose to render and which ones we choose not to render.  Further, the decisions we make to render something act as a feedback loop and reinforces the underlying tagging.  If we encounter an unusual tagging system, it is better to not render it, and have it be missing from the map, until it can be shown that the tagging system enjoys a high enough level of data consumer support such that our decision to add render support would not inappropriately reinforce the unusual scheme.  I would certainly not want to be responsible for causing an otherwise fungible tagging scheme to become entrenched because of a render decision on the part of this project.

Germany is hardly the only location where we are faced with these decisions; there are a number of unusual and one-off route networks in the US that present us with a similar decision point, for example in #119, as well as other cases in Texas and New York.  Highway shield rendering based on route relations is still in its infancy in OSM, and as a new renderer in this space, we are proceeding forward cautiously when deciding what to render.

If this means that the rendering displayed in some parts of the map is slightly degraded while we take an appropriate length of time to carefully consider the ramifications of adding render support, so be it.  

Since we do not presently render shields for Germany (although there is a PR considering adding such support), the proposed approach simply results the status quo - a generic white box with the ref drawn inside.

**Reply by zekefarwell** [reply]
Please also realize that the primary objective of the style is to provide a more familar map of North America for North American mappers, just as the [German style](https://wiki.openstreetmap.org/wiki/German_Style) does for German mappers.  We will also add support for highway shields in other areas of the world where it makes sense, but North America is the focus.  If things don't look quite right in other areas, that is OK.

**Reply by 1ec5** [reply]
> The wiki clearly states: "In **some** countries, network values for highway routes follow a hierarchical scheme."

I [introduced](https://wiki.openstreetmap.org/wiki/Special:Diff/1123692) this wording in order to distinguish between the `<country>:<type>` “basic” scheme and the more complex “hierarchical” scheme, both of which are supported by this renderer. As far as I can tell, it has always been the case that “most” countries prefix their networks with the country code, with Germany being an outlier in this regard. (It’s a different story for public transportation and recreational route networks, but the [extreme overloading](https://wiki.openstreetmap.org/wiki/Homonymous_keys#Differing_formats) of `network` is a topic for another day…)

**Comment by adamfranco** [comment]
I've done a little research via [taginfo](https://taginfo.openstreetmap.org/keys/network#values) to determine the scope of `network=*` usage as relates to the road networks in question.

I began with a non-exhaustive sample of values identified that match the `network=<country_code>:*` prefixed format and found this format applied to more than 100,000 relations in just these 12 countries:
* 34,339 `network=RU:*`
* 20,000+ `network=US:*`
* 10,882 `network=UA:*`
* 6,779 `network=CZ:*`
* 6,485 `network=RO:*`
* 4,656 `network=JP:*`
* 4,500+ `network=CA:*`
* 3,500+ `network=IN:*`
* 2,741 `network=BY:*`
* 2,200+ `network=IT:*`
* 2,021 `network=PY:*`
* 2,000+ `network=BR:*`

In contrast, the values previously identified that don't match the `network=<country_code>:*` format are applied to only 252 relations total in 3 countries:
* 37 `network=motorway` (Switzerland)
* 179 `network=BAB` (Germany)
* 36 `network=*高速道路` (Japan)

Based on these usage numbers, I think we can safely say that country-prefixed `network` values are the clear norm globally and exceptions to this format are very small. Given the lack of tagging consensus in the German community around `network=BAB` and the preponderance of prefixed-usage in Japan already I feel comfortable not supporting un-prefixed `route=road` `network=*` values for national/sub-national routes.
